cql3/statements/select_statement.cc

```diff
@@ -404,26 +405,52 @@ select_statement::do_execute(query_processor& qp,
 
     auto key_ranges = _restrictions->get_partition_key_ranges(options);
 
+    bool send_tablet = false;
+
+    auto token = dht::token();
+    unsigned shard;
+
+    if (key_ranges.size() == 1 && query::is_single_partition(key_ranges.front())) {
+        token = key_ranges[0].start()->value().as_decorated_key().token();
+        shard = _schema->table().shard_of(token);
+    }
+
+    auto&& table = qp.proxy().local_db().find_column_family(_schema);
+    auto erm = table.get_effective_replication_map();
+    locator::tablet_replica_set tablet_replicas = {};
+    auto token_range = std::pair<dht::token, dht::token>();
+
```

Comment:
I'm sure storage_proxy is performing this check for us, perhaps we can get the result back from it instead of redoing it here.

Reply:
My initial idea was to somehow extract tablet info from storage_proxy, but @tgrabiec advised me against it and encouraged me to do it the way I did here

Reply:
Well, this will have a performance impact, so I encourage @tgrabiec to reconsider his advice

Reply:
It's theoretically possible, but much harder to plug into the code in storage_proxy. We're crossing many layers to get to storage proxy in some cases, including query_pager and index reads. We need to propagate info in both directions, to storage_proxy, whether the driver expects the info for this query, and back, the info itself. 

The performance impact could be reduced if we have erm->check_locality(), then in the happy path it wouldn't allocate anything, just perform lookups in tablet vector and host_id/shard comparisons. find_column_family() can be avoided with s->table().

Reply:
You are right, logging show that the query is going through the right path, it is trying to get information about tablets, but the replicas are always empty, but it's good news I suppose that performance doesn't suffer in such a situation.

Reply:
What do you mean that the replicas are empty?

Reply:
@sylwiaszunejko What's the command line? When running with many shards (e.g. `-c4`) I would expect some shard mismatches. Also note that by default tables are not tablet-based, would be good to test this case too.

Reply:
@avikivity I mean the tables are not tablet based @tgrabiec  I wasn't using additional options, that was my mistake, e.g. with `-c4` there are about 370 extra instructions

Reply:
Well we should work to reduce it. 

Please check tasks and allocations too, likely a task/allocation was added.